text/_util.ts

```diff
@@ -1,9 +1,18 @@
 // Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
 
+const CAPITALIZED_WORD_REGEXP = /\p{Lu}\p{Ll}+/u; // e.g. Apple
+const ACRONYM_REGEXP = /\p{Lu}+(?=(\p{Lu}\p{Ll})|\P{L}|\b)/u; // e.g. ID, URL, also handles an acronym followed by a capitalized word like in HTMLDivElement
+const LOWERCASED_WORD_REGEXP = /(\p{Ll}+)/u; // e.g. apple
+const ANY_LETTERS = /\p{L}+/u; // will also match any sequence of letters in languages without a concept of upper/lower case
+const DIGITS_REGEXP = /\p{N}+/u; // e.g. 123
+
+const WORD_OR_NUMBER_REGEXP = new RegExp(
+  `${CAPITALIZED_WORD_REGEXP.source}|${ACRONYM_REGEXP.source}|${LOWERCASED_WORD_REGEXP.source}|${ANY_LETTERS.source}|${DIGITS_REGEXP.source}`,
+  "gu",
+);
+
 export function splitToWords(input: string) {
-  input = input.replaceAll(/[^a-zA-Z0-9\s-_]/g, "");
-  if (/[\s-_]+/.test(input)) return input.split(/[\s-_]+/);
-  return input.split(/(?=[A-Z])+/);
+  return input.match(WORD_OR_NUMBER_REGEXP) || [];
```

Comment:
@guy-borderless Can you give example that this regexp doesn't match? Is `[]` appropriate fallback in that case?

Reply:
"no words" may be an expected happy path response so why not allow it to fit in the normal path of the code, if it's logical. I think an empty array here is ok and less annoying than null. We can do null though. Anyway, added a test for that https://github.com/guy-borderless/deno_std/blob/dfd42c715e36fda7a94de287ecf99dd14fe11ba4/text/_util_test.ts#L14

Reply:
Thanks! Looks nice test cases to illustrate what happens to non-alphanumeric